matrix/Dual.hpp

```diff
@@ -0,0 +1,361 @@
+Dual<Scalar, N> ceil(const Dual<Scalar, N>& a)
+{
+    return Dual<Scalar, N>(ceil(a.value));
+}
+
+// floor
+template <typename Scalar, size_t N>
+Dual<Scalar, N> floor(const Dual<Scalar, N>& a)
+{
+    return Dual<Scalar, N>(floor(a.value));
+}
+
+// fmod
+template <typename Scalar, size_t N>
+Dual<Scalar, N> fmod(const Dual<Scalar, N>& a, Scalar mod)
+{
+    return Dual<Scalar, N>(a.value - Scalar(size_t(a.value / mod)) * mod, a.derivative);
+}
+
+// max
+template <typename Scalar, size_t N>
+Dual<Scalar, N> max(const Dual<Scalar, N>& a, const Dual<Scalar, N>& b)
+{
+    return a.value >= b.value ? a : b;
+}
+
+// min
+template <typename Scalar, size_t N>
+Dual<Scalar, N> min(const Dual<Scalar, N>& a, const Dual<Scalar, N>& b)
+{
+    return a.value < b.value ? a : b;
+}
+
+// isnan
+template <typename Scalar, size_t N>
+bool isnan(const Dual<Scalar, N>& a)
+{
+    return isnan(a.value);
+}
+
+// isfinite
+template <typename Scalar, size_t N>
+bool isfinite(const Dual<Scalar, N>& a)
+{
+    return isfinite(a.value);
+}
+
+// isinf
+template <typename Scalar, size_t N>
+bool isinf(const Dual<Scalar, N>& a)
+{
+    return isinf(a.value);
+}
+
+// trig
+
+// sin
+template <typename Scalar, size_t N>
+Dual<Scalar, N> sin(const Dual<Scalar, N>& a)
+{
+    return Dual<Scalar, N>(sin(a.value), cos(a.value) * a.derivative);
+}
+
+// cos
+template <typename Scalar, size_t N>
+Dual<Scalar, N> cos(const Dual<Scalar, N>& a)
+{
+    return Dual<Scalar, N>(cos(a.value), -sin(a.value) * a.derivative);
+}
+
+// tan
+template <typename Scalar, size_t N>
+Dual<Scalar, N> tan(const Dual<Scalar, N>& a)
+{
+    Scalar real = tan(a.value);
+    return Dual<Scalar, N>(real, (Scalar(1) + real * real) * a.derivative);
+}
+
+// asin
+template <typename Scalar, size_t N>
+Dual<Scalar, N> asin(const Dual<Scalar, N>& a)
+{
+    Scalar asin_d = Scalar(1) / sqrt(Scalar(1) - a.value * a.value);
+    return Dual<Scalar, N>(asin(a.value), asin_d * a.derivative);
+}
+
+// acos
+template <typename Scalar, size_t N>
+Dual<Scalar, N> acos(const Dual<Scalar, N>& a)
+{
+    Scalar acos_d = -Scalar(1) / sqrt(Scalar(1) - a.value * a.value);
+    return Dual<Scalar, N>(acos(a.value), acos_d * a.derivative);
+}
+
+// atan
+template <typename Scalar, size_t N>
+Dual<Scalar, N> atan(const Dual<Scalar, N>& a)
+{
+    Scalar atan_d = Scalar(1) / (Scalar(1) + a.value * a.value);
+    return Dual<Scalar, N>(atan(a.value), atan_d * a.derivative);
+}
+
+// atan2
+template <typename Scalar, size_t N>
+Dual<Scalar, N> atan2(const Dual<Scalar, N>& a, const Dual<Scalar, N>& b)
+{
+    // derivative is equal to that of atan(a/b), so substitute a/b into atan and simplify
+    Scalar atan_d = Scalar(1) / (a.value * a.value + b.value * b.value);
+    return Dual<Scalar, N>(atan2(a.value, b.value), (a.derivative * b.value - a.value * b.derivative) * atan_d);
+}
+
+template <typename Scalar, size_t M, size_t N>
+Matrix<Scalar, M, N> derivative(const Matrix<Dual<Scalar, N>, M, 1>& input)
```

Comment:
This function has the same name as one of the member variables. Maybe we can chance this function name to something like "getDerivative"